Add logging to serializer exceptions

joelhoisko opened the pull request:
This helps us with having a clearer log mesage about which event couldn't get properly serialized.

Not quite sure that should we just do a `throw` after the logging or is there some fancier system for us to more gracefully tell what went wrong?

woksin commented on Source/Events/EventConverter.cs:
```diff
@@ -73,28 +96,36 @@ public Contracts.UncommittedAggregateEvents ToProtobuf(UncommittedAggregateEvent
-                    source.ExecutionContext.ToExecutionContext(),
-                    @event);
-            }
+                var artifact = ToSDK(source.Type);
+                try
+                {
+                    var @event = _serializer.JsonToEvent(artifact, source.Content);
+                    if (source.External)
+                    {
+                        return new CommittedExternalEvent(
+                            source.EventLogSequenceNumber,
+                            source.Occurred.ToDateTimeOffset(),
+                            source.EventSourceId.To<EventSourceId>(),
+                            source.ExecutionContext.ToExecutionContext(),
+                            source.ExternalEventLogSequenceNumber,
+                            source.ExternalEventReceived.ToDateTimeOffset(),
+                            @event);
+                    }
+                    else
+                    {
+                        return new CommittedEvent(
+                            source.EventLogSequenceNumber,
+                            source.Occurred.ToDateTimeOffset(),
+                            source.EventSourceId.To<EventSourceId>(),
+                            source.ExecutionContext.ToExecutionContext(),
+                            @event);
+                    }
+                }
+                catch (UnableToInstantiateInstanceOfType ex)
+                {
+                    _logger.Warning(ex, "Couldn't deserialize artifact: {Artifact} to a committed event.", artifact);
+                    throw;
+                }
```
There are a couple of things here I'm thinking of. 
* The variable `artifact`, it's not really artifact, it is the `eventType`. I know it's not your code but would be nice to change it here :) 
* We could have a little bit more developer friendly log message by using this format instead `Couldn't deserialize artifact '{Artifact}' to '{EventType}' event` where `Artifact` is `source.Type.Id.To<ArtifactId>`
* We should throw an exception here, but I don't think that we should throw the original one. Maybe we should have a `CouldNotDeserializeEvent` that takes an `ArtifactId` and the `Type` of the `Event`? And also I think we can drop giving the exception `ex` as an argument to `_logger.Warning`

woksin replied:
@jakhog  opinions? As you have more experience actually using this. What's your take?

jakhog replied:
First of all, I would change the caught exception type to just `Exception` since more things can go wrong here (JSON syntax errors e.g.), and in stead of logging create the exception that you mention @woksin, but also send in the actual JSON content of the event, and its position in the event store (scope and sequence number).

woksin replied:
That’s a good point with the either throw or log. We should have some guidelines around this down the line.

Also agree on the JSON content part

woksin commented on Source/Events/EventConverter.cs:
```diff
@@ -109,16 +140,24 @@ public CommittedAggregateEvents ToSDK(Contracts.CommittedAggregateEvents source)
-                    aggregateRootVersion++,
-                    eventSource.ExecutionContext.ToExecutionContext(),
-                    @event);
+                    var artifact = ToSDK(eventSource.Type);
+                    try
+                    {
+                        var @event = _serializer.JsonToEvent(artifact, eventSource.Content);
+                        return new CommittedAggregateEvent(
+                            eventSource.EventLogSequenceNumber,
+                            eventSource.Occurred.ToDateTimeOffset(),
+                            source.EventSourceId.To<EventSourceId>(),
+                            aggregateRoot,
+                            aggregateRootVersion++,
+                            eventSource.ExecutionContext.ToExecutionContext(),
+                            @event);
+                    }
+                    catch (UnableToInstantiateInstanceOfType ex)
+                    {
+                        _logger.Warning(ex, "Couldn't deserialize artifact {Artifact} to CommittedAggregateEvent.", artifact);
+                        throw;
+                    }
```
Same as above

jakhog replied:
Same follow-up as above.

jakhog commented:
Just to get my head back into the problem, lets re-iterate the problem so we agree. As far as I remember, the cause was that the definition of an event had changed causing an exception when the head was told to process that by the runtime. The effect was quite clear in terms of an exception shown in the head-logs, but it was very hard to track down the actual instance of the event (we had to look through all of them). Does that sound about right?

I think the initial goal was to add more context to the error, so not just the standard unmarshalling exception, but also the JSON content of the event, and the scope and event log sequence number so that finding it in the event store would be easier.

An interesting question pops up - which you also kind of asked @joelhoisko - should we log + throw? I think that we should in general only do one of the two, and in this specific case we definitely have to throw. So I don't think we need to write log messages here.

My suggestion would be that we create a new exception `CouldNotDeserializeEvent` that takes in the `artifact`, `content`, `scope`, `eventLogSequenceNumber` and the original `exception` to make a nicer message that eventually ends up in the head logs. And that we don't do explicit logging here as well, as it will most likely result in duplicate log entries.

Does that make sense?

jakhog commented on Source/Events/EventConverter.cs:
```diff
@@ -18,29 +19,43 @@ public class EventConverter : IEventConverter
-                Public = IsPublicEvent(@event.Event),
-                Content = _serializer.EventToJson(@event.Event),
-            };
+                try
+                {
+                    return new Contracts.UncommittedEvent
+                    {
+                        Artifact = ToProtobuf(@event.Event.GetType()),
+                        EventSourceId = @event.EventSource.ToProtobuf(),
+                        Public = IsPublicEvent(@event.Event),
+                        Content = _serializer.EventToJson(@event.Event),
+                    };
+                }
+                catch (UnableToInstantiateInstanceOfType ex)
+                {
+                    _logger.Warning(ex, "Couldn't serialize uncommitted event: {Type} to Json.", @event.Event.GetType());
+                    throw;
+                }
+            }
```
There is nothing wrong with this code as it is rethrowing the exception, but I don't think we need to do anything special here. That specific exception should never be thrown by this code, and as @woksin said, this is a different kind of problem than what we need to solve right now.

jakhog commented on Source/Events/EventConverter.cs:
```diff
@@ -59,12 +74,20 @@ public Contracts.UncommittedAggregateEvents ToProtobuf(UncommittedAggregateEvent
+                {
+                    events.Events.Add(new Contracts.UncommittedAggregateEvents.Types.UncommittedAggregateEvent
+                    {
+                        Artifact = ToProtobuf(@event.GetType()),
+                        Public = IsPublicEvent(@event),
+                        Content = _serializer.EventToJson(@event),
+                    });
+                }
+                catch (UnableToInstantiateInstanceOfType ex)
                 {
-                    Artifact = ToProtobuf(@event.GetType()),
-                    Public = IsPublicEvent(@event),
-                    Content = _serializer.EventToJson(@event),
-                });
+                    _logger.Warning(ex, "Couldn't serialize uncommitted aggregate event: {Type} to Json.", @event.GetType());
+                    throw;
+                }
```
Same follow-up as above.

woksin commented:
> Just to get my head back into the problem, lets re-iterate the problem so we agree. As far as I remember, the cause was that the definition of an event had changed causing an exception when the head was told to process that by the runtime. The effect was quite clear in terms of an exception shown in the head-logs, but it was very hard to track down the actual instance of the event (we had to look through all of them). Does that sound about right?
> 
That sounds about right not me

joelhoisko commented:
Added more exceptions now so that we can properly add the scope to log messages. It's a bit messy but the `try/catch` logic in `AbstractFilterProcessor`/EventStore/EventHandlerProcessor` forced my hand a bit here. So you basically get 3 exceptions rolled into one with this approach.

Here's how it looks like (I added an extra `"}"` at the end of the JSON content string to produce this exception):

![image](https://user-images.githubusercontent.com/10163775/88278091-8391b780-ccea-11ea-8979-2b72d6adb366.png)

woksin commented on Source/Events/CouldNotDeserializeEvent.cs:
```diff
@@ -0,0 +1,32 @@
+            ArtifactId artifact,
+            Type eventType,
+            string content,
+            ulong sequenceNumber,
```
I think that we should use the Concepts here
```suggestion
            EventLogSequenceNumber sequenceNumber,
```

joelhoisko replied:
The value given to the exception comes from the `Contracts.CommittedEvent.EventLogSequenceNumber` which is a `ulong`, so it doesn't have any `To<EventLogSequenceNumber>()` functions like the `Uuid`'s.

woksin replied:
Yes, you’re right

woksin commented on Source/Events.Filters/Internal/AbstractFilterProcessor.cs:
```diff
@@ -57,13 +58,20 @@ protected override RetryProcessingState GetRetryProcessingState(FilterEventReque
+            }
+            catch (CouldNotDeserializeEvent ex)
+            {
+                throw new CouldNotDeserializeEventFromScope(request.ScopeId.ToGuid(), ex);
```
Should pass in the actual ScopeId Concept
```suggestion
                throw new CouldNotDeserializeEventFromScope(request.ScopeId.To<ScopeId>(), ex);
```

woksin commented on Source/Events.Handling/Internal/EventHandlerProcessor.cs:
```diff
@@ -135,29 +135,35 @@ protected override async Task<EventHandlerResponse> Handle(HandleEventRequest re
             }
+            catch (CouldNotDeserializeEvent ex)
+            {
+                throw new CouldNotDeserializeEventFromScope(request.Event.ScopeId.ToGuid(), ex);
```
Same as above.
```suggestion
                throw new CouldNotDeserializeEventFromScope(request.Event.ScopeId.To<ScopeId>(), ex);
```

woksin commented:
> Added more exceptions now so that we can properly add the scope to log messages. It's a bit messy but the `try/catch` logic in `AbstractFilterProcessor`/EventStore/EventHandlerProcessor` forced my hand a bit here. So you basically get 3 exceptions rolled into one with this approach.
>

I think this is close to as good as we can get it now without changing the interfaces. In the future I would like the IEventConverter to return something like a conversion result saying whether or not it failed, or something like that

woksin commented:
Maybe delay this for when we release the barebones SDK? @joelhoisko

woksin commented:
I'll pull it in without releasing
